src/lib.rs

```diff
@@ -0,0 +1,489 @@
+        _ => return false,
+    };
+}
+
+/// Store a trace (or "trace pending" marker in the trace cache for the specified location.
+fn set_trace(loc: JitLocation, trace: JitTrace) {
+    // Sanity checks.
+    {
+        // Extra scope for dropping the lock prior to the write() lock later (avoids deadlock).
+        let trace_map = TRACES.read().unwrap();
+        let old_val = trace_map.get(&loc);
+        if trace == JitTrace::TracePending {
+            // If we are collecting a trace, then the location should not already be in the trace cache.
+            assert!(old_val == None);
+        } else if trace == JitTrace::FinalisedTrace {
+            // If we are storing a finished trace, then it should have been marked as "pending" first.
+            assert!(old_val == Some(&JitTrace::TracePending));
+        }
+    }
+
+    TRACES.write().unwrap().insert(loc, trace);
+}
+
+/// Shared hot counts.
+///
+/// Maps a location to its hot count. Increments of the counter are atomic due to the RwLock.
+lazy_static! {
+    static ref HOT_COUNTS: RwLock<HashMap<JitLocation, HotCount>> = RwLock::new(HashMap::new());
+}
+
+/// Get the hot count for the given location.
+fn hot_count(loc: JitLocation) -> HotCount {
+    let ret = match HOT_COUNTS.read().unwrap().get(&loc) {
+        Some(v) => *v,
+        None => 0,
+    };
+    ret
+}
+
+/// Atomically increment and return the hot count for the given location.
+fn incr_hot_count(loc: JitLocation) -> HotCount {
+    // XXX don't store locations which cannot be loop heads.
+    let mut counts = HOT_COUNTS.write().unwrap();
+    let loc_count = counts.entry(loc).or_insert(0);
+    *loc_count += 1;
+    *loc_count
+}
+
+/// Thread-specific meta-tracer context.
+///
+/// Each thread has one of instance of this inside a thread-local.
+#[derive(Debug)]
+pub struct JitThreadContext {
+    phase: JitPhase,
+    // XXX PT packet buffer goes here.
+}
+
+impl JitThreadContext {
+    pub fn new() -> Self {
+        Self {
+            phase: JitPhase::Interpreting,
+        }
+    }
+
+    /// Resets the per-thread JIT state to defaults.
+    #[cfg(test)]
+    fn reset(&mut self) {
+        self.phase = JitPhase::Interpreting;
+    }
+
+    /// Get the current thread's JIT phase.
+    fn current_phase() -> JitPhase {
+        let phase = JIT_THREAD_CX.with(|k| {
+            k.borrow().phase.clone()
+        });
+        phase
+    }
+
+    /// Transition to a new JIT phase.
+    fn phase_transition(phase: JitPhase) {
+        JIT_THREAD_CX.with(|k| {
+            k.borrow_mut().phase = phase;
+        });
+    }
+
+    /// Compile a trace from the raw PT packets (in parallel with the interpreter)..
+    ///
+    // XXX eventually we probably want a thread pool, as having an unbounded number of threads
+    // is probably bad.
+    fn compile_trace(loc: JitLocation) {
+        let name = format!("compile_loc_{}", loc);
+        debug!("new compilation thread: {}", name);
```

Comment:
Isn't this just the same as `debug!("new blah: compile_loc_{}", loc)`?

Reply:
`name` is used again below, so I wanted to store it.

Reply:
OK.